### PR TITLE
Revise the code guide in the contributor guide

### LINKED ISSUE #1214: Discuss in code development guide when to use and not use function aliases
I just happened across #1101, which proposes and implements a change to limit aliases to `plasmapy.formulary` and other very commonly known aliases, such as some very commonly used ones in Langmuir analysis world.  The reasoning behind that is well worth recording in our development guide, so I suggest summarizing that in a bullet point in `docs/development/code_guide.rst`.

Related to #1117.

### PR DESCRIPTION
**Pull request by namurphy**
This PR is going to be an update of the code guide that follows up on #1156, #1163, and #1198.  This falls under #1117.  

I'm starting it now since I wanted to make sure to talk about making comments future-proof (inspired by [this comment](https://github.com/PlasmaPy/PlasmaPy/pull/1299#issuecomment-938266090)), but I'm guessing I won't finish it in 2021 😬 .

This might be a good PR in which to address #582, #685, #797, #1227, and #1234...but those might wait until another PR (especially #1227).  We will definitely need to talk about `validate_quantities` and `particle_input` here too.

Closes #1214.

### CODE REVIEW

**Comment by namurphy** [comment]
And a reminder to myself...  I'll probably be making a bunch of changes to `docs/common_links.rst` which is a frequently edited file.  To start out with I should put all of the newly defined links at the bottom of `code_guide.rst`, and then move them to `docs/common_links.rst` later on in order to avoid dreaded merge conflicts of joy and despair.

**Comment by namurphy** [comment]
- [ ] (Briefly) describe the different techniques use to optimize code.  (Maybe a different PR, but if so, raise an issue instead.)

**Comment by rocco8773** [comment]
How to import functionality from `plasmapy.formulary`?

1. For end users, they should import directly from `plasmapy.formulary` (e.g. `from plasmapy.formulary import plasma_frequency`), because this would lead to more robustness against changing internal pkg structure.
2. For contributors developing code,..
    1. For locations outside `plasmapy.formulary` functionality can be imported from `plasmapy.formulary`.
    2. For locations inside `plasmapy.formulary`, ...
       * Importing from `plasmapy.formulary` will absolutely cause circular imports, so this option can not be done.
       * Functionality to be directly imported from its sub-module or the sub-module itself should be imported.
       * Importing functionality from a sub-module can still cause circular imports, in that case import the sub-module and dot access the functionality.  For example, 
           ```python
           from plasmapy.formulary import frequencies

           wp = frequencies.plasmapy_frequency(...)
           ```

**Comment by namurphy** [comment]
- [ ] Describe how to handle deprecations and changes in keyword arguments.
 - [ ] Describe how to know if a change is breaking or not.  ← might go in changelog guide, so perhaps create a separate issue

**Comment by namurphy** [comment]
Add that we can use tools like [WebPlotDigitizer](https://automeris.io/WebPlotDigitizer/) to get values from graphs.  Was used in #1189.

**Comment by namurphy** [comment]
- [ ] Specify the dependency tree for subpackages.  `utils` can't import from anywhere else; `particles` can import from `utils`, `formulary` can import from `particles` & `utils`, etc.

File: docs/development/code_guide.rst [hunk 1]
```diff
@@ -1,305 +1,444 @@
...
+  - `itertools` — Functions creating iterators for efficient looping
+  - `json` — JSON encoder and decoder
+  - `numbers` — Numeric abstract base classes, often used for type hint annotations
+  - `os` — Miscellaneous operating system interfaces
+  - `pdb` — The Python debugger
+  - `re` — Regular expression operations
+  - `string` — Common string operations
+  - `sys` — System-specific parameters and functions
+  - `typing` — Support for type hints
+  - `warnings` — Warning control
```
[the remark this answers is not in the document]
Comment (namurphy) [reply docs/development/code_guide.rst:52]: Yeah, I was going back and forth on this one.  My main motivation for including this is that I've been slow to learn about and appreciate everything that's in the STL.  I think I'll drop the list, but include a link to the STL and say that it's worth learning more about.

**Comment by namurphy** [comment]
- [ ] Make sure to incorporate the following text, which came from `docs/formulary/index.rst`:

> Notes for developers
> ====================
> 
> .. This section should be moved to the contributor guide.
> 
> Values should be returned as an Astropy |Quantity| in SI units.
> 
> If a quantity has several names, then the function name should be the
> one that provides the most physical insight into what the quantity
> represents.  For example, ``gyrofrequency`` indicates gyration, while
> Larmor frequency indicates that this frequency is somehow related to a
> human (or perhaps a cat?) named Larmor.  Similarly, using ``omega_ce``
> as a function name for this quantity will make the code less readable to
> people who are unfamiliar with the notation or use a different symbol.
> 
> The docstrings for plasma parameter methods should describe the
> physics associated with these quantities in ways that are
> understandable to students who are taking their first course in plasma
> physics while still being useful to experienced plasma physicists.

**Comment by namurphy** [comment]
- [ ] Mention annotations and what they're used for
 - [ ] Bring up how to use `validate_quantities` and `particle_input`

**Comment by namurphy** [comment]
Thinking about this more, it's probably a better idea to break this up into multiple, smaller pull requests.

**Comment by namurphy** [comment]
Closing this in favor of a few separate pull requests (see above).  The two main sections that I haven't moved over are on:

 - Python standard library
 - IDEs

The paragraph on IDEs could perhaps be shortened and put in #1656.  Right now I'm not planning on putting the section on the Python standard library anywhere else, but it might be worth thinking about again in shorter form if we add a resources page to the contributor guide.

**Pull request closed without merging**